runtime/near-vm-logic/src/logic.rs

```diff
@@ -300,27 +308,38 @@ impl<'a> VMLogic<'a> {
         String::from_utf8(buf).map_err(|_| HostError::BadUTF8.into())
     }
 
-    /// Helper function to read UTF-16 assembly script formatted string from guest memory.
+    /// Helper function to read UTF-16 formatted string from guest memory.
     /// # Errors
     ///
     /// * If string extends outside the memory of the guest with `MemoryAccessViolation`;
     /// * If string is not UTF-16 returns `BadUtf16`.
-    fn get_assembly_script_utf16_string(&mut self, ptr: u64) -> Result<String> {
-        if ptr < 4 {
-            return Err(HostError::BadUTF16.into());
-        }
-        let mut slice = [0u8; 4];
-        let buf = Self::memory_get(self.memory, ptr - 4, 4)?;
-        slice.copy_from_slice(&buf);
-        let len: u32 = u32::from_le_bytes(slice);
-        if len % 2 != 0 {
-            return Err(HostError::BadUTF16.into());
-        }
-        let buffer = Self::memory_get(self.memory, ptr, u64::from(len))?;
+    fn get_utf16_string(&mut self, len: u64, ptr: u64) -> Result<String> {
+        let input: Vec<u8>;
         let mut u16_buffer = Vec::new();
-        for i in 0..((len / 2) as usize) {
-            let c = u16::from(buffer[i * 2]) + u16::from(buffer[i * 2 + 1]) * 0x100;
-            u16_buffer.push(c);
+        let max_len = self.config.max_log_len;
+        if len != std::u64::MAX {
+            input = Self::memory_get(self.memory, ptr, len as u64)?;
+            if len % 2 != 0 || len > max_len {
+                return Err(HostError::BadUTF16.into());
+            }
+            for i in 0..((len / 2) as usize) {
+                let hi = input[i as usize * 2 + 1] as u16 * 0x100;
+                let lo = input[i as usize * 2] as u16;
+                u16_buffer.push(hi + lo);
+            }
+        } else {
+            for i in 0..=(max_len / 2) {
+                Self::try_fit_mem(self.memory, ptr, i)?;
```

Comment:
need to check `i * 2 + 2` here

Reply:
What do you mean check `i*2+2`? `try_fit_mem` already checks the boundaries.

Reply:
So we can't check `ptr + i` here cause it might overflow u64. Instead we use a hack to check length. I guess we should be able to do
`Self::try_fit_mem(self.memory, ptr + i, 2)?;` since we'll iteratively check for u64 overflows

Reply:
`Self::try_fit_mem(self.memory, ptr + i, 2)?` is bad, because user can provide `ptr == u64::MAX`.
The only check that I see working here is:
`Self::try_fit_mem(self.memory, ptr, i*2+size_of::<u16>())?`, since `i` is guaranteed to be small, less than `self.config.max_log_len`.

Reply:
Sorry, I meant to use `ptr + i * size_of::<u16>()`
Like here:
`Self::try_fit_mem(self.memory, ptr + i * size_of::<u16>(), size_of::<u16>())?`

Because we start with `i == 0`, there is no overflow for the initial check. But we'll check the sum for `i == 1` doesn't overflow. So it's safe

Reply:
Ok